### ISSUE TITLE
Xamarin.Google.Android.DataTransport.TransportRuntime uses a bad version of Xamarin.Google.Dagger

### ISSUE THREAD

**Issue by bramborman** [xamarin/XamarinComponents#1069]
In our Xamarin.Android project I just updated `Xamarin.Firebase.Messaging` package to the latest available version `121.0.1` which depends on `Xamarin.Google.Android.DataTransport.TransportRuntime >= 2.2.4`. Version `2.2.4` does not exist so it's resolved to version `2.2.5` which depends on `Xamarin.Google.Dagger >=2.13.0`. That is a problem however as the app crashes on startup with the following error:

    Java.Lang.NoClassDefFoundError: 'Failed resolution of: Lcom/google/android/datatransport/runtime/dagger/internal/Factory;'

This issue can be easily resolved by updating `Xamarin.Google.Dagger` to version `>= 2.25.2.1` but it should not be required to do it manually - `Xamarin.Google.Android.DataTransport.TransportRuntime` should depend on the correct version IMO.

**Comment by michaldobrodenka** [comment]
I had similar problem:

https://stackoverflow.com/questions/66071588/xamarin-android-app-crash-on-start-with-firebase-after-upgrading-xamarin-azure-n

**Comment by moljac** [comment]
Thanks for feedback.

It uses the version that was the latest at the time bindings were done.

closing this one

**Comment by bramborman** [comment]
> It uses the version that was the latest at the time bindings were done.

@moljac ok but it's really a problem that should be solved as the app crashes when developers try to use the library... The package should be updated to use the correct version so developers can safely use it without spending a lot of time on finding an issue they should not even face.

**Comment by moljac** [comment]
@bramborman 

I agree, but there is tons of feedback in AndroidX and GPS-FB repos for the same problem.

All 3 repos are open sourced... Do you get it where I am heading?

Otherwise it will be updated someday. Depending on other tasks...

**Comment by bramborman** [comment]
@moljac

> I agree, but there is tons of feedback in AndroidX and GPS-FB repos for the same problem.

So it's a broadly faced issue which should prioritize the fix.

> All 3 repos are open sourced... Do you get it where I am heading?

I know they're open sourced, but I also thought Xamarin is supported by Microsoft as Microsoft wants developers to use their platform... The readme says this project is _created by Xamarin and others in the community_ which probably means it's rather a community project, than a fully supported product I guess? However AndroidX is, correct me if I'm wrong, a fully supported one as it's the base for modern Android app development and I'd expect Microsoft to fix such a severe issue rather than relying on the community...

**Comment by michaldobrodenka** [comment]
It's a MS/Xamarin problem. I've spent many hours figuring out why my app keeps crashing on startup after updating to AndroidX. The problem is just bug in Nuget Xamarin.Firebase.Messaging. Without this issue and info from @bramborman I would probably gave up AndroidX migration and blame Xamarin platform as buggy.

**Comment by moljac** [comment]
@bramborman 

TL&DR - PRs are welcome

> However AndroidX is, correct me if I'm wrong, a fully supported one as it's the base for modern Android app development and I'd expect Microsoft to fix such a severe issue rather than relying on the community...

I must correct you.

See:

*  https://github.com/xamarin/AndroidX/pulls?q=is%3Apr+author%3ACheesebaron
*  https://github.com/xamarin/AndroidX/pulls?q=is%3Apr+author%3Agfinol
*  https://github.com/xamarin/AndroidX/pulls?q=is%3Apr+author%3AChasakisD

This users are not Microsoft/Xamarin. One of them works on MVVMCross and MVVMCross != Microsoft/Xamarin.

So not even AndroidX is "fully" supported by Microsoft/Xamarin. Yes we approve PRs and control contributions from users just like we control contributions from each of us. Just to prevent huge mistakes and be in sync with future plans.

Now check this:

*   https://github.com/xamarin/XamarinComponents/pulls?q=is%3Apr+author%3Aantonioseric

There is discussion how much to maintain. For Xamarin.Forms we need several AndroidX packages and Maps from GPS-FB. I fight hard to keep "important" stuff from google at least. Important is matter of definition/discussion. I added MLKit recently. There are requests for `androidx.security`, `androidx.window`etc...

So when I see:

> It's a MS/Xamarin problem.

I say: not only. And go further to solve other "priority" issues. Whereby "priority" is also matter of definition/discussion.

> and I'd expect Microsoft to fix such a severe issue rather than relying on the community...

It is not relying, but if users have their own priorities - they can do something about it.

Regarding bugs - yes they are present, but some users know how to report them and even help pinpoint them, while others report them by mistake. See:

https://github.com/xamarin/AndroidX/issues/170

At least they are decent and admit they made mistake.

There are alternatives to Xamarin, but Xamarin ecosystem is bigger due to bindings. There is no Flutter or React Native support for ALL MLKit packages for example. Xamarin has it.

And there are users that talk about bugs and keep on talking...

Now the decision is yours. I will (or someone else) will update those packages soon.

**Comment by moljac** [comment]
@michaldobrodenka 

Search in GPS-FB would reveal several workarounds for Dagger:

*   https://github.com/xamarin/GooglePlayServicesComponents/issues?q=is%3Aissue+Dagger

Crashlytics: https://github.com/xamarin/GooglePlayServicesComponents/issues/393

Cast: https://github.com/xamarin/GooglePlayServicesComponents/issues/431

All include manual addition of Xamarin.Dagger nuget package.

The reason is concept called "transitive dependencies". To be precise in dependency chain some nuget points to older dependency which uses older artifacts.

`Xamarin.Azure.NotificationHubs.Android` uses Firebase which uses Dagger which uses older Datatransport.
Adding latest Dagger will pull in correct/newer Datatransport.

So one could blame `Xamarin.Azure.NotificationHubs.Android`, but it is not that simple.

**Comment by bramborman** [comment]
@moljac

>So not even AndroidX is "fully" supported by Microsoft/Xamarin.

Not sure if you got what I mean... I mean support similar to, for example .NET (Core), which is also open source, accepts new features (after some review of course), fixes and improvements from the community, but still it gets a full Microsoft support in the way that if there is a bug that breaks something, they fix it and don't just say _you can fix it yourself, we are working on something else at this time_. Don't get me wrong, from what I understood, there are not many maintainers (which is in my opinion the root problem here, something Microsoft should improve imho), and those who are around are, as I see from what you say, overflown with tasks such as new features, existing bugs, etc. so they (you) have to prioritize. The thing that got me into this discussion, and what surprised me the most, is that this issue got closed with a message that sounds kinda like _we don't care, we won't fix it_. I'd expect the issue would get tagged as duplicate and closed or let open until it's fixed 😊

>There are alternatives to Xamarin, but Xamarin ecosystem is bigger due to bindings.

That's why I like Xamarin - it's not _reimplementing the platform_, it brings it to your language. That's exactly why, despite all the issues, it's still the platform of choice for me. Just if it got more full-time devs😅 And considering how big the platforms are and how many official libraries are there, I truly am impressed with the work around it you people do, because Xamarin is I believe a really complex piece of tech and so I understand the delays of new features, such as bindings updates to newer versions etc. 😊

**Comment by moljac** [comment]
@bramborman 

> Not sure if you got what I mean...

OK. I might have misunderstood something.

> similar to, for example .NET (Core), which is also open source, accepts new features (after some review of course), fixes and improvements from the community, but still it gets a full Microsoft support in the way that if there is a bug that breaks something, they fix it

The closest to the .NET Core would be Xamarin.Android, Xamarin.IOS, Xamarin.Mac as cornerstones for Xamarin.Forms and MAUI in the future. AndroidX could be cut down to several packages including only one from GPS-FB - Maps with few dependencies if exist. All together it would be 30 packages +/- few. And there were plans to cut down. So, for the community it would be good to get involved - in order to protect their investment in know-how (remember Xamarin.Insights or HockeyApp?). 

> don't just say you can fix it yourself, we are working on something else at this time.

If the fix is changing few version numbers why demanding from us to switch context? We might be fixing another bug which is important for someone else.

See:

*   https://github.com/xamarin/XamarinComponents/pull/1068/files
*   https://github.com/xamarin/XamarinComponents/pull/1071/files
*   https://github.com/xamarin/XamarinComponents/pull/1062/files

I think .NET Core features and bugfixes are much more complex... If users get stuck they can mention us and we do help if necessary.

> there are not many maintainers (which is in my opinion the root problem here, something Microsoft should improve imho)

True, but depending on how long this pandemic will last + effects of upcoming recession I would not bet on the watching that movie (increasing number of devs).

> and those who are around are, as I see from what you say, overflown with tasks such as new features, existing bugs, etc. so they (you) have to prioritize. 

Another true fact.

> what surprised me the most, is that this issue got closed with a message that sounds kinda like we don't care, we won't fix it. 

Not completely true. It got closed because there is workaround -manually adding nuget. I planned to switch to XamarinComponents repo next week, but that might be postponed, because I spent whole morning pinpointing bug with community member. 

>  issue would get tagged as duplicate and closed or let open until it's fixed

It is not duplicate and if there is comment providing workaround I tend to close them. Otherwise just browsing the open issues and trying to repro them with growing number of issues will kill me/us.

> Just if it got more full-time devs

It is our dream too. And not only that - once upon the time there was team that worked on components AKA plugins in order to provide xplat ecosystem (xplat API for any SDK).

> because Xamarin is I believe a really complex piece of tech and so I understand the delays of new features, such as bindings updates to newer versions

It is very complex and a lot of people do not understand that. Our tooling is gtting better and better, but other platforms evolve and grow too... We really try to reduce the delays, but that is difficult.

All I wanted to point out is - with little help of community - life for all of us would be easier and nicer...

Bottom line - updates are coming soon.
